core/src/main/java/org/apache/iceberg/Partition.java

```diff
@@ -0,0 +1,392 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.Objects;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.types.Types;
+
+public class Partition implements IndexedRecord {
+  private PartitionData partitionData;
+  private int specId;
+  private long dataRecordCount;
+  private int dataFileCount;
+  private long dataFileSizeInBytes;
+  // optional fields are kept as objects instead of primitive.
+  private Long posDeleteRecordCount;
+  private Integer posDeleteFileCount;
+  private Long eqDeleteRecordCount;
+  private Integer eqDeleteFileCount;
+  // Commit time of snapshot that last updated this partition
+  private Long lastUpdatedAt;
+  // ID of snapshot that last updated this partition
+  private Long lastUpdatedSnapshotId;
+
+  public enum Column {
+    PARTITION_DATA,
+    SPEC_ID,
+    DATA_RECORD_COUNT,
+    DATA_FILE_COUNT,
+    DATA_FILE_SIZE_IN_BYTES,
+    POSITION_DELETE_RECORD_COUNT,
+    POSITION_DELETE_FILE_COUNT,
+    EQUALITY_DELETE_RECORD_COUNT,
+    EQUALITY_DELETE_FILE_COUNT,
+    LAST_UPDATED_AT,
+    LAST_UPDATED_SNAPSHOT_ID
+  }
+
+  public Partition() {}
+
+  public Partition(StructLike key, Types.StructType keyType) {
+    this.partitionData = toPartitionData(key, keyType);
+    this.specId = 0;
+    this.dataRecordCount = 0L;
+    this.dataFileCount = 0;
+    this.dataFileSizeInBytes = 0L;
+    this.posDeleteRecordCount = 0L;
+    this.posDeleteFileCount = 0;
+    this.eqDeleteRecordCount = 0L;
+    this.eqDeleteFileCount = 0;
+  }
+
+  public static PartitionBuilder builder() {
+    return new PartitionBuilder();
+  }
+
+  public PartitionData partitionData() {
+    return partitionData;
+  }
+
+  public int specId() {
+    return specId;
+  }
+
+  public long dataRecordCount() {
+    return dataRecordCount;
+  }
+
+  public int dataFileCount() {
+    return dataFileCount;
+  }
+
+  public long dataFileSizeInBytes() {
+    return dataFileSizeInBytes;
+  }
+
+  public Long posDeleteRecordCount() {
+    return posDeleteRecordCount;
+  }
+
+  public Integer posDeleteFileCount() {
+    return posDeleteFileCount;
+  }
+
+  public Long eqDeleteRecordCount() {
+    return eqDeleteRecordCount;
+  }
+
+  public Integer eqDeleteFileCount() {
+    return eqDeleteFileCount;
+  }
+
+  public Long lastUpdatedAt() {
+    return lastUpdatedAt;
+  }
+
+  public Long lastUpdatedSnapshotId() {
+    return lastUpdatedSnapshotId;
+  }
+
+  synchronized void update(ContentFile<?> file, Snapshot snapshot) {
```

Comment:
Nit: why the change to make it synchronized, are we expecting to update the same Partition reference from multiple threads?

Reply:
It depends on engine integration. Some engine like Dremio, writes manifest files and datafiles concurrently for a single snapshot operations. In that case we need a synchronization as there can be multiple data files belong to same partition.

Reply:
Sure, I guess I was thinking that only after writing the data files (even if the data files are written concurrently) would we update the partition stats (serially) but as you said this is really an engine implementation detail.